File: evals/lib/vectors.ts
/**
 * Shared by precompute.ts (write) and run-retrieval.ts (read): a readFileSync Buffer can be
 * backed by Node's shared pool with a nonzero byteOffset, so the offset must survive the
 * Buffer->Float32Array reinterpretation. Kept in one place so both sides can't drift apart.
 */
export function packVectors(vectors: number[][], dim: number): Buffer {
  const flat = new Float32Array(vectors.length * dim)
  vectors.forEach((v, i) => flat.set(v, i * dim))
  return Buffer.from(flat.buffer)
}

export function unpackVectors(raw: Buffer, dim: number, count: number): Float32Array[] {
  const flat = new Float32Array(raw.buffer.slice(raw.byteOffset, raw.byteOffset + raw.byteLength))
  return Array.from({ length: count }, (_, i) => flat.subarray(i * dim, (i + 1) * dim))
}

File: evals/lib/vectors.py
from collections.abc import Sequence

import numpy as np
from numpy.typing import NDArray

Vectors = NDArray[np.float32]


def pack_vectors(vectors: Sequence[Sequence[float]], dim: int) -> bytes:
    flat = np.zeros((len(vectors), dim), dtype=np.float32)
    for i, v in enumerate(vectors):
        flat[i, : len(v)] = v
    return flat.tobytes()


def unpack_vectors(raw: bytes, dim: int, count: int) -> Vectors:
    return np.frombuffer(raw, dtype=np.float32, count=dim * count).reshape(count, dim)
